File: src/back-end/FunChess/src/FunChess.Core.Chess/Cell.cs
using System.Diagnostics.CodeAnalysis;
using System.Runtime.InteropServices;
using FunChess.Core.Chess.Enums;

namespace FunChess.Core.Chess;

public sealed class Cell
{
    private static readonly Dictionary<byte, Cell> CellsVariations = new();
    
    static Cell()
    {
        CellsVariations.Add(0, new Cell(null, null));
        for (int i = 1; i <= 2; i++)
        {
            for (int j = 4; j <= 128; j <<= 1)
            {
                CellsVariations.Add((byte)(j + i), new Cell((Piece)j, (Team)i));
            }
        }
        Empty = CellsVariations[0];
    }
    
    private Cell(Piece? piece, Team? team)
    {
        Piece = piece;
        Team = team;
    }
    
    public Piece? Piece { get; private init; }
    
    public Team? Team { get; private init; }
    
    [MemberNotNullWhen(true, nameof(Piece), nameof(Team))]
    public bool IsFromTeam(Team team)
    {
        return Team == team;
    }
    
    [MemberNotNullWhen(false, nameof(Piece), nameof(Team))]
    public bool IsEmpty()
    {
        return this == Empty;
    }

    public static explicit operator byte(Cell cell)
    {
        if (cell.IsEmpty()) return 0;
        return (byte)(cell.Team + (byte)cell.Piece);
    }

    public static explicit operator Cell(byte value)
    {
        if (value == 0) return Empty;

        var piece = (Piece)value;
        var team = (Team)(value - (byte)piece);
        
        return Get(piece, team);
    }

    public static readonly Cell Empty;

    public static Cell Get(Piece piece, Team team)
    {
        byte value = (byte)((byte)piece + team);
        _ = CellsVariations.TryGetValue(value, out Cell? cell);

        return cell!;
    }
}

File: src/front-end/fun-chess-app/src/core/chess/Cell.ts
import Piece from "./Piece";
import Team from "./enums/Team";
import Bishop from "./pieces/Bishop";
import King from "./pieces/King";
import Knight from "./pieces/Knight";
import Pawn from "./pieces/Pawn";
import Queen from "./pieces/Queen";
import Rook from "./pieces/Rook";

class Cell {
    private static readonly CellsVariations = new Map<number, Cell>();

    public static readonly Empty = new Cell(null, null);

    static {
        for (let i = 1; i <= 2; i++) {
            Cell.CellsVariations.set(i + 4, new Cell(King.instance, i));
            Cell.CellsVariations.set(i + 8, new Cell(Queen.instance, i));
            Cell.CellsVariations.set(i + 16, new Cell(Rook.instance, i));
            Cell.CellsVariations.set(i + 32, new Cell(Knight.instance, i));
            Cell.CellsVariations.set(i + 64, new Cell(Bishop.instance, i));
            Cell.CellsVariations.set(i + 128, new Cell(Pawn.instance, i));
        }
    }

    private constructor(piece: Piece | null, team: Team | null) {
        this.piece = piece;
        this.team = team;
    }

    public readonly piece: Piece | null;

    public readonly team: Team | null;

    public isFromTeam(team: Team): boolean {
        return this.team === team
    }

    public isEmpty(): boolean {
        return this === Cell.Empty;
    }

    public toByte(): number {
        if (this.isEmpty()) return 0;
        return Piece.toByte(this.piece) + this.team!;
    }

    public static parse(value: number): Cell {
        if (value === 0) return this.Empty;

        const piece = Piece.parse(value);
        const team: Team = value - Piece.toByte(piece);

        return Cell.get(piece, team);
    }

    public static get(piece: Piece, team: Team): Cell {
        const value = Piece.toByte(piece) + team;
        return Cell.CellsVariations.get(value)!;
    }
}

export default Cell;